crates/bevy_ptr/README.md

```diff
@@ -6,8 +6,104 @@
 
-The `bevy_ptr` crate provides low-level abstractions for working with pointers in a more safe way than using rust's raw pointers.
+Pointers in computer programming, according to Wikipedia, are "objects in many programming languages that stores a memory address". 
+They're a fundamental building block for constructing more complex data structures.
 
-Rust has lifetimed and typed references (`&'a T`), unlifetimed and typed references (`*const T`), but no lifetimed but untyped references.
-`bevy_ptr` adds them, called `Ptr<'a>`, `PtrMut<'a>` and `OwningPtr<'a>`.
-These types are lifetime-checked so can never lead to problems like use-after-frees and must always point to valid data.
+They're also *the* definitive source of memory safety bugs. You can dereference a null pointer. Access a pointer after the underlying 
+memory has been freed. Ignore type safety and misread or mutate the underlying memory improperly.
+
+Rust is a programming language that heavily relies on its types to enforce correctness, and by proxy, memory safety. As a result, 
+Rust has an entire zoo of types for working with pointers, and a graph of safe and unsafe conversions to among them to make working 
+with them safer.
+
+`bevy_ptr` is a crate that attempts to bridge the gap between the full blown unsafety of `*mut ()` and the safe `&'a T`, allowing users 
+to build progressively to choose what invariants to uphold.
+
+## How to build a Borrow (from scratch)
+Correctly and safety converting a pointer into a valid borrow is at the core of all `unsafe` code in Rust. Looking at the documentation for
+[`(*const T)::as_ref`], a pointer must satisfy *all* of the following conditions:
+
+  * The pointer must be properly aligned.
+  * The pointer cannot be null, even for zero sized types.
+  * The pointer must be within bounds of a valid allocated object (on the stack or the heap).
+  * The pointer must point to an initialized instance of `T`.
+  * The newly assigned lifetime should be valid for the value that the pointer is targeting.
+  * The code must enforce Rust's aliasing rules. Only one mutable borrow or arbitrarily many read-only borrows may exist to a value at any given moment
+    in time, and converting from `&T` to `&mut T` is always not allowed.
+
+Note these rules aren't final and are still in flux as the Rust Project hashes out what exactly are the pointer aliasing rules, but the expectation is that the
+final set of constraints are going to be a superset of this list, not a subset.
+
+This list already is non-trivial to satisfy in isolation. Thankfully, the Rust core/standard library provides a progressive list of pointer types that help
+build these safety guarantees...
+
+## Standard Pointers
+
+|Pointer Type       |Lifetime'ed|Mutable|Strongly Typed|Aligned|Not Null|Forbids Aliasing|Forbids Arithmetic|
+|:------------------|:----------|:------|:-------------|:------|:-------|:---------------|:-----------------|
```

Comment:
What is your reasoning for including the colon before the dashes? (`|:----|`) Do tables in other modules use this syntax instead of just `|---|`?

Reply:
It makes the alignment explicit. Probably could be removed.

Reply:
I like specifying alignment in these tables :)